### PR TITLE
#14141 SI project export and post-processing handler

### PR DESCRIPTION
**Pull request by michael-jong-ds**
1. Added handler to update database after reading
2. Delete successfully exported projects (soft delete)
3. Send request to endpoint after export

### CODE REVIEW

**Comment by michael-jong-ds** [comment]
updated kak @tonoman3g, please review, thanks

File: src/handlers/rex-export.handler.ts [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+  file_page_size: 999,
+  original_filename: 'si.pdf',
+};
+
+export async function handleExport(configFile: string, data: DataPayload) {
+  // TODO: implement post processing and delete the exported file
+  await handleExportProjects(configFile, { unzip: true });
+
+  // receives json format with reading_result and document_data
+  await handleSaveToDatabase(configFile, data.id);
```
Comment (tonoman3g) [inline src/handlers/rex-export.handler.ts:22]: if you finish the post processing, please call the functions here @yosua1011

File: src/handlers/save-to-database.handler.ts [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+    let documentData = {};
+    let readingResult = {};
+
+    getLogger().info(`processing export results of project ${folder}...`);
+    readdirSync(folder).forEach(async (file) => {
+      getLogger().info(`checking file ${file}...`);
+      const json = readJSONFile(resolve(folder, file));
+
+      // file name should be <ID>_<PAGE_NUMBER>.jpg
+      const filename = path.parse(file).name;
```
Comment (tonoman3g) [inline src/handlers/save-to-database.handler.ts:32]: belom diimplement karena nungguin yosua?

File: src/handlers/save-to-database.handler.ts [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+import path, { resolve } from 'path';
+import { getConfig, setConfigByJSONFile } from '../config/config';
+import { getDatabaseValidators } from '../config/schema/validator';
+import { initDatabase } from '../database';
+import { Team15 } from '../database/entities/teamPayloads/team_15.entity';
+import { getRepository } from '../database/repository';
+import { getLogger } from '../logger';
+import { readJSONFile } from '../utils/readJSONFile';
+
+export async function handleSaveToDatabase(configFile: string, id: number) {
```
Comment (tonoman3g) [inline src/handlers/save-to-database.handler.ts:11]: apakah ini handler?

handler itu adalah command yang diexecute kan, kalau ini bukan handler, bikin folder baru aja di dalam src @michael-jong-ds

File: src/handlers/export-projects.handler.ts [hunk 4]
```diff
@@ -94,8 +95,22 @@ async function _handleExportProjects(configFile: string, { unzip }: { unzip: boo
   getLogger().info(`found ${projectsToExport.length} projects to export`);
 
   const results = await runProjectExport(projectsToExport, unzip, scriptState);
-  await checkProjectExportJobs(results);
+  const successResults = await checkProjectExportJobs(results);
   await scriptState.save();
+
+  const successProjectNames = successResults.map((exportResult) => exportResult.projectName);
```
Comment (tonoman3g) [inline src/handlers/export-projects.handler.ts:101]: tambahin parameter di handler ini aja, seperti unzip

misal `{ deleteProjectAfterExport: true }`

File: src/handlers/export-projects.handler.ts [hunk 5]
```diff
@@ -94,8 +95,22 @@ async function _handleExportProjects(configFile: string, { unzip }: { unzip: boo
...
+  const projectsToDelete = Array.from(scriptState.getTeamProjectsState().getProjects()).filter(
```
Comment (tonoman3g) [inline src/handlers/export-projects.handler.ts:102]: hooo jadi harus iterate semua project ya 🤔 

bisa pakai project id yang ada di meta-data.json gak?

kan skg di zip project, uda ada meta-data.json yang isinya projectId, jadi kita bisa pakai projectId tersebut.

File: src/handlers/send-request.handler.ts [hunk 6]
```diff
@@ -0,0 +1,64 @@
+import axios from 'axios';
+import { setConfigByJSONFile } from '../config/config';
+import { getDatabaseValidators } from '../config/schema/validator';
+import { initDatabase } from '../database';
+import { Team15 } from '../database/entities/teamPayloads/team_15.entity';
+import { getRepository } from '../database/repository';
+import { getLogger } from '../logger';
+
+export async function sendRequestToEndpoint(configFile: string, id: number) {
```
Comment (tonoman3g) [inline src/handlers/send-request.handler.ts:9]: apakah ini handler?

handler itu adalah command yang diexecute kan, kalau ini bukan handler, bikin folder baru aja di dalam src @michael-jong-ds

File: src/handlers/send-request.handler.ts [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+  initDatabase();
+  const team15Repository = await getRepository(Team15);
+  const data = await team15Repository.findOneByOrFail(Number(id));
+  const exportEndpoint = process.env.EXPORT_ENDPOINT;
+
+  const payload = {
+    document_data: data.document_data,
+    document_path: data.hcp_ori_document_dir,
+    transaction_id: data.id,
+    status: isOCRSuccessful(data.ocr_status) ? 'Success' : 'Failed',
```
Comment (tonoman3g) [inline src/handlers/send-request.handler.ts:20]: 'Success' and 'Failed' should be moved to constant OCR_STATUS_SUCCESS, OCR_STATUS_FAILED

File: src/handlers/send-request.handler.ts [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+
+  let counterRetry = 0;
+  const LIMIT_RETRY = Number(process.env.LIMIT_RETRY);
+  while (counterRetry < LIMIT_RETRY) {
+    try {
+      const response = await axios({
+        method: 'POST',
+        headers: { 'content-type': 'application/x-www-form-urlencoded' },
+        data: payload,
+        url: exportEndpoint,
```
Comment (tonoman3g) [inline src/handlers/send-request.handler.ts:39]: endpointnya ga ada autheticationnya ya?

File: src/handlers/send-request.handler.ts [hunk 9]
```diff
@@ -0,0 +1,64 @@
...
+        timeout: 10000,
```
Comment (tonoman3g) [inline src/handlers/send-request.handler.ts:40]: timeout 10 detik cukup kah? apa 30 detik aja?

File: src/handlers/send-request.handler.ts [hunk 10]
```diff
@@ -0,0 +1,64 @@
...
+    received_request: data.received_request,
+    start_ocr: data.start_ocr,
+    end_ocr: data.end_ocr,
+    continuous_index: data.continuous_index,
+  };
+
+  let counterRetry = 0;
+  const LIMIT_RETRY = Number(process.env.LIMIT_RETRY);
+  while (counterRetry < LIMIT_RETRY) {
+    try {
```
Comment (tonoman3g) [inline src/handlers/send-request.handler.ts:34]: add sleep around 1-2s before retrying

File: src/handlers/send-request.handler.ts [hunk 11]
```diff
@@ -0,0 +1,64 @@
...
+          message: error.message,
+        });
+        counterRetry += 1;
+      }
+    }
+  }
+}
+
+function isOCRSuccessful(ocr_status) {
+  return ocr_status === 'Read';
```
Comment (tonoman3g) [inline src/handlers/send-request.handler.ts:63]: Read juga dimove ke constant OCR_STATUS_READ

File: src/handlers/save-to-database.handler.ts (hunk 2 above)
Comment (michael-jong-ds) [reply src/handlers/save-to-database.handler.ts:32]: expected result dari input harusnya sudah bener <ID>_<PAGE_NUMBER>.jpg

File: src/handlers/export-projects.handler.ts (hunk 5 above)
Comment (michael-jong-ds) [reply src/handlers/export-projects.handler.ts:102]: oh, mksdnya project-data.json?
bisa, tpi I figured exportnya bakal `unzip: true`, jdi ak bikin gini
should I change it jdi ngambil dri json itu?

File: src/handlers/send-request.handler.ts (hunk 8 above)
Comment (michael-jong-ds) [reply src/handlers/send-request.handler.ts:39]: from [here](https://docs.google.com/document/d/1SKrWnjDaK5XWCUXlGSSZFzRF8QQx3utRb9h19kRgvdQ/edit#) kyknya gaada

File: src/export/post-process.ts [hunk 12]
```diff
@@ -0,0 +1,18 @@
...
+
+export async function postProcessDocumentData(data) {
+  const payload = {
+    document_data: data,
+  };
+  const postProcessEndpoint = process.env.POST_PROCESS_DOCUMENT_DATA_ENDPOINT;
+
+  const response = await axios({
+    method: 'POST',
+    headers: { 'content-type': 'application/x-www-form-urlencoded' },
```
Comment (tonoman3g) [inline src/export/post-process.ts:11]: the header should not be application/x-www-form-urlencoded, it should be application/json

Comment (michael-jong-ds) [reply src/export/post-process.ts]: updated

File: src/handlers/export-projects.handler.ts [hunk 13]
```diff
@@ -94,8 +102,25 @@ async function _handleExportProjects(configFile: string, { unzip }: { unzip: boo
...
+    const projectsToDelete = Array.from(scriptState.getTeamProjectsState().getProjects()).filter(
+      ([_name, projectState]) => {
+        if (successProjectNames.includes(projectState.projectName)) {
+          return true;
+        }
+        return false;
+      },
+    );
+
+    for (const [_name, projectState] of projectsToDelete) {
```
Comment (tonoman3g) [inline src/handlers/export-projects.handler.ts:120]: oh ya make the deletion more verbose @michael-jong-ds by logging it here

File: src/export/send-request.ts [hunk 14]
```diff
@@ -0,0 +1,67 @@
...
+    received_request: data.received_request,
+    start_ocr: data.start_ocr,
+    end_ocr: data.end_ocr,
+    continuous_index: data.continuous_index,
+  };
+
+  let counterRetry = 0;
+  const LIMIT_RETRY = Number(process.env.LIMIT_RETRY);
+  while (counterRetry < LIMIT_RETRY) {
+    try {
```
Comment (tonoman3g) [inline src/export/send-request.ts:36]: also please log the payload here

File: src/export/send-request.ts [hunk 15]
```diff
@@ -0,0 +1,67 @@
...
+      const response = await axios({
+        method: 'POST',
+        headers: { 'content-type': 'application/json' },
+        data: payload,
+        url: exportEndpoint,
+        timeout: 30000,
+      });
+      counterRetry += LIMIT_RETRY;
+      return response;
```
Comment (tonoman3g) [inline src/export/send-request.ts:45]: if success, log the success here
```

logger.info(...., {responseBody: response.body})
return response
```

**Comment by michael-jong-ds** [comment]
kak @tonoman3g 

Logs screenshot 
Screenshot from 2022-11-09 16-49-53
Screenshot from 2022-11-09 16-50-13
Screenshot from 2022-11-09 16-50-06

mongodb express
Screenshot from 2022-11-09 16-52-07

**Review by tonoman3g — APPROVED** [review]

**Pull request merged**